File: packages/contracts-ts/src/offline-entitlement.ts
import { createPublicKey, verify } from 'node:crypto';

import { controlPlaneDocumentValidator } from './generated/index.js';

export const OFFLINE_ENTITLEMENT_VALIDITY_SECONDS = 604_800 as const;

export const OfflineEntitlementVerdict = {
  Verified: 'verified',
  OnlineVerificationRequired: 'online-verification-required',
} as const;

export type OfflineEntitlementVerdict =
  (typeof OfflineEntitlementVerdict)[keyof typeof OfflineEntitlementVerdict];

export type OfflineEntitlementSigningKeyStatus = 'current' | 'previous' | 'retired';

export interface OfflineEntitlementSigningKey {
  readonly keyId: string;
  readonly publicKeyBytes: string;
  readonly status: OfflineEntitlementSigningKeyStatus;
  readonly notBeforeUnixSeconds: number;
  readonly notAfterUnixSeconds: number;
}

export interface OfflineEntitlementVerificationContext {
  readonly accountId: string;
  readonly deviceBinding: string;
  readonly audience: string;
  readonly entitlementVersion: number;
  readonly nowUnixSeconds: number;
}

export interface TrustedTimeStore {
  readLastTrustedUnixSeconds(): number | undefined;
  writeLastTrustedUnixSeconds(value: number): void;
}

interface EnvelopeView {
  readonly payloadBytes: string;
  readonly signature: string;
  readonly keyId: string;
  readonly audience: string;
  readonly deviceBinding: string;
  readonly issuedAt: string;
  readonly expiresAt: string;
}

export interface OfflineEntitlementClaims {
  readonly schemaVersion: '1.0';
  readonly accountId: string;
  readonly deviceBinding: string;
  readonly audience: string;
  readonly entitlementVersion: number;
  readonly issuedAt: string;
  readonly expiresAt: string;
  readonly validitySeconds: typeof OFFLINE_ENTITLEMENT_VALIDITY_SECONDS;
}

export const encodeOfflineEntitlementPayload = (claims: OfflineEntitlementClaims): Buffer =>
  Buffer.from(JSON.stringify(claims), 'utf8');

const ED25519_SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');
const CANONICAL_UTC = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.000Z$/u;
const BASE64URL = /^[A-Za-z0-9_-]+={0,2}$/u;

const asRecord = (value: unknown): Readonly<Record<string, unknown>> | undefined =>
  typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Readonly<Record<string, unknown>>)
    : undefined;

const stringProperty = (
  record: Readonly<Record<string, unknown>>,
  property: string,
): string | undefined => {
  const value = record[property];
  return typeof value === 'string' ? value : undefined;
};

const envelopeView = (value: unknown): EnvelopeView | undefined => {
  const record = asRecord(value);
  if (record === undefined) return undefined;

  const payloadBytes = stringProperty(record, 'payloadBytes');
  const signature = stringProperty(record, 'signature');
  const keyId = stringProperty(record, 'keyId');
  const audience = stringProperty(record, 'audience');
  const deviceBinding = stringProperty(record, 'deviceBinding');
  const issuedAt = stringProperty(record, 'issuedAt');
  const expiresAt = stringProperty(record, 'expiresAt');

  if (
    payloadBytes === undefined ||
    signature === undefined ||
    keyId === undefined ||
    audience === undefined ||
    deviceBinding === undefined ||
    issuedAt === undefined ||
    expiresAt === undefined
  ) {
    return undefined;
  }

  return { payloadBytes, signature, keyId, audience, deviceBinding, issuedAt, expiresAt };
};

const decodeBase64Url = (value: string): Buffer | undefined => {
  if (!BASE64URL.test(value) || value.length % 4 === 1) return undefined;

  const decoded = Buffer.from(value, 'base64url');
  if (decoded.toString('base64url') !== value.replace(/=+$/u, '')) return undefined;
  return decoded;
};

const parseCanonicalUtcSeconds = (value: string): number | undefined => {
  if (!CANONICAL_UTC.test(value)) return undefined;
  const milliseconds = Date.parse(value);
  if (!Number.isFinite(milliseconds) || new Date(milliseconds).toISOString() !== value) {
    return undefined;
  }
  return milliseconds / 1_000;
};

export const decodeOfflineEntitlementPayload = (
  payloadBytes: Buffer,
): OfflineEntitlementClaims | undefined => {
  let value: unknown;
  try {
    value = JSON.parse(payloadBytes.toString('utf8')) as unknown;
  } catch {
    return undefined;
  }

  const record = asRecord(value);
  if (record === undefined || Object.keys(record).length !== 8) return undefined;

  const schemaVersion = stringProperty(record, 'schemaVersion');
  const accountId = stringProperty(record, 'accountId');
  const deviceBinding = stringProperty(record, 'deviceBinding');
  const audience = stringProperty(record, 'audience');
  const issuedAt = stringProperty(record, 'issuedAt');
  const expiresAt = stringProperty(record, 'expiresAt');
  const entitlementVersion = record['entitlementVersion'];
  const validitySeconds = record['validitySeconds'];

  if (
    schemaVersion === undefined ||
    accountId === undefined ||
    deviceBinding === undefined ||
    audience === undefined ||
    issuedAt === undefined ||
    expiresAt === undefined ||
    !Number.isSafeInteger(entitlementVersion) ||
    !Number.isSafeInteger(validitySeconds)
  ) {
    return undefined;
  }

  if (schemaVersion !== '1.0' || validitySeconds !== OFFLINE_ENTITLEMENT_VALIDITY_SECONDS) {
    return undefined;
  }

  return {
    schemaVersion,
    accountId,
    deviceBinding,
    audience,
    entitlementVersion: entitlementVersion as number,
    issuedAt,
    expiresAt,
    validitySeconds,
  };
};

const verificationRequired = (): OfflineEntitlementVerdict =>
  OfflineEntitlementVerdict.OnlineVerificationRequired;

const verifyCandidate = (
  input: unknown,
  keyRing: readonly OfflineEntitlementSigningKey[],
  context: OfflineEntitlementVerificationContext,
  trustedTimeStore: TrustedTimeStore,
): OfflineEntitlementVerdict => {
  const envelope = envelopeView(input);
  if (envelope === undefined) return verificationRequired();

  const signingKey = keyRing.find(({ keyId }) => keyId === envelope.keyId);
  if (signingKey === undefined || signingKey.status === 'retired') {
    return verificationRequired();
  }

  const payloadBytes = decodeBase64Url(envelope.payloadBytes);
  const signatureBytes = decodeBase64Url(envelope.signature);
  const publicKeyBytes = decodeBase64Url(signingKey.publicKeyBytes);
  if (
    payloadBytes === undefined ||
    signatureBytes?.length !== 64 ||
    publicKeyBytes?.length !== 32
  ) {
    return verificationRequired();
  }

  const publicKey = createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, publicKeyBytes]),
    format: 'der',
    type: 'spki',
  });
  if (!verify(null, payloadBytes, publicKey, signatureBytes)) {
    return verificationRequired();
  }

  if (!controlPlaneDocumentValidator(input)) return verificationRequired();

  const claims = decodeOfflineEntitlementPayload(payloadBytes);
  if (claims === undefined) return verificationRequired();

  const issuedAt = parseCanonicalUtcSeconds(claims.issuedAt);
  const expiresAt = parseCanonicalUtcSeconds(claims.expiresAt);
  const envelopeIssuedAt = parseCanonicalUtcSeconds(envelope.issuedAt);
  const envelopeExpiresAt = parseCanonicalUtcSeconds(envelope.expiresAt);
  if (
    issuedAt === undefined ||
    expiresAt === undefined ||
    envelopeIssuedAt === undefined ||
    envelopeExpiresAt === undefined
  ) {
    return verificationRequired();
  }

  const lastTrustedTime = trustedTimeStore.readLastTrustedUnixSeconds();
  if (
    !Number.isSafeInteger(context.nowUnixSeconds) ||
    (lastTrustedTime !== undefined &&
      (!Number.isSafeInteger(lastTrustedTime) || context.nowUnixSeconds < lastTrustedTime)) ||
    context.nowUnixSeconds < issuedAt ||
    context.nowUnixSeconds > expiresAt ||
    expiresAt - issuedAt !== OFFLINE_ENTITLEMENT_VALIDITY_SECONDS ||
    claims.validitySeconds !== OFFLINE_ENTITLEMENT_VALIDITY_SECONDS ||
    issuedAt !== envelopeIssuedAt ||
    expiresAt !== envelopeExpiresAt ||
    claims.schemaVersion !== '1.0' ||
    claims.accountId !== context.accountId ||
    claims.deviceBinding !== context.deviceBinding ||
    claims.deviceBinding !== envelope.deviceBinding ||
    claims.audience !== context.audience ||
    claims.audience !== envelope.audience ||
    claims.entitlementVersion !== context.entitlementVersion ||
    issuedAt < signingKey.notBeforeUnixSeconds ||
    issuedAt > signingKey.notAfterUnixSeconds ||
    context.nowUnixSeconds < signingKey.notBeforeUnixSeconds ||
    context.nowUnixSeconds > signingKey.notAfterUnixSeconds
  ) {
    return verificationRequired();
  }

  trustedTimeStore.writeLastTrustedUnixSeconds(context.nowUnixSeconds);
  return OfflineEntitlementVerdict.Verified;
};

export const verifyOfflineEntitlementBytes = (
  input: unknown,
  keyRing: readonly OfflineEntitlementSigningKey[],
  context: OfflineEntitlementVerificationContext,
  trustedTimeStore: TrustedTimeStore,
): OfflineEntitlementVerdict => {
  try {
    return verifyCandidate(input, keyRing, context, trustedTimeStore);
  } catch {
    return verificationRequired();
  }
};

File: apps/desktop/src-tauri/src/offline_entitlement.rs
use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};
use ed25519_dalek::{Signature, VerifyingKey};
use liiiraa_contracts_rust::validate_control_plane_document;
use serde::Deserialize;
use serde_json::Value;

pub const OFFLINE_ENTITLEMENT_VALIDITY_SECONDS: i64 = 604_800;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum OfflineEntitlementVerdict {
    Verified,
    OnlineVerificationRequired,
}

impl OfflineEntitlementVerdict {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Verified => "verified",
            Self::OnlineVerificationRequired => "online-verification-required",
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum OfflineEntitlementSigningKeyStatus {
    Current,
    Previous,
    Retired,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct OfflineEntitlementSigningKey {
    pub key_id: String,
    pub public_key_bytes: [u8; 32],
    pub status: OfflineEntitlementSigningKeyStatus,
    pub not_before_unix_seconds: i64,
    pub not_after_unix_seconds: i64,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct OfflineEntitlementVerificationContext<'a> {
    pub account_id: &'a str,
    pub device_binding: &'a str,
    pub audience: &'a str,
    pub entitlement_version: u64,
    pub now_unix_seconds: i64,
}

pub trait TrustedTimeStore {
    fn read_last_trusted_unix_seconds(&self) -> Option<i64>;
    fn write_last_trusted_unix_seconds(&mut self, value: i64) -> Result<(), ()>;
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct OfflineEntitlementClaims {
    schema_version: String,
    account_id: String,
    device_binding: String,
    audience: String,
    entitlement_version: u64,
    issued_at: String,
    expires_at: String,
    validity_seconds: i64,
}

struct EnvelopeView<'a> {
    payload_bytes: &'a str,
    signature: &'a str,
    key_id: &'a str,
    audience: &'a str,
    device_binding: &'a str,
    issued_at: &'a str,
    expires_at: &'a str,
}

fn string_property<'a>(input: &'a Value, property: &str) -> Option<&'a str> {
    input.as_object()?.get(property)?.as_str()
}

fn envelope_view(input: &Value) -> Option<EnvelopeView<'_>> {
    Some(EnvelopeView {
        payload_bytes: string_property(input, "payloadBytes")?,
        signature: string_property(input, "signature")?,
        key_id: string_property(input, "keyId")?,
        audience: string_property(input, "audience")?,
        device_binding: string_property(input, "deviceBinding")?,
        issued_at: string_property(input, "issuedAt")?,
        expires_at: string_property(input, "expiresAt")?,
    })
}

fn decode_base64url(value: &str) -> Option<Vec<u8>> {
    if value.is_empty()
        || value.len() % 4 == 1
        || !value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'-'))
    {
        return None;
    }

    let decoded = URL_SAFE_NO_PAD.decode(value).ok()?;
    (URL_SAFE_NO_PAD.encode(&decoded) == value).then_some(decoded)
}

fn leap_year(year: i64) -> bool {
    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
}

fn days_in_month(year: i64, month: i64) -> Option<i64> {
    match month {
        1 | 3 | 5 | 7 | 8 | 10 | 12 => Some(31),
        4 | 6 | 9 | 11 => Some(30),
        2 if leap_year(year) => Some(29),
        2 => Some(28),
        _ => None,
    }
}

fn decimal(input: &[u8]) -> Option<i64> {
    input.iter().try_fold(0_i64, |value, byte| {
        if byte.is_ascii_digit() {
            Some(value * 10 + i64::from(byte - b'0'))
        } else {
            None
        }
    })
}

pub(crate) fn parse_canonical_utc_seconds(value: &str) -> Option<i64> {
    let bytes = value.as_bytes();
    if bytes.len() != 24
        || bytes[4] != b'-'
        || bytes[7] != b'-'
        || bytes[10] != b'T'
        || bytes[13] != b':'
        || bytes[16] != b':'
        || &bytes[19..24] != b".000Z"
    {
        return None;
    }

    let year = decimal(&bytes[0..4])?;
    let month = decimal(&bytes[5..7])?;
    let day = decimal(&bytes[8..10])?;
    let hour = decimal(&bytes[11..13])?;
    let minute = decimal(&bytes[14..16])?;
    let second = decimal(&bytes[17..19])?;
    if year < 1970
        || day == 0
        || day > days_in_month(year, month)?
        || hour > 23
        || minute > 59
        || second > 59
    {
        return None;
    }

    let adjusted_year = year - i64::from(month <= 2);
    let era = adjusted_year.div_euclid(400);
    let year_of_era = adjusted_year - era * 400;
    let adjusted_month = month + if month > 2 { -3 } else { 9 };
    let day_of_year = (153 * adjusted_month + 2) / 5 + day - 1;
    let day_of_era = year_of_era * 365 + year_of_era / 4 - year_of_era / 100 + day_of_year;
    let days_since_epoch = era * 146_097 + day_of_era - 719_468;

    days_since_epoch
        .checked_mul(86_400)?
        .checked_add(hour * 3_600 + minute * 60 + second)
}

fn verification_required() -> OfflineEntitlementVerdict {
    OfflineEntitlementVerdict::OnlineVerificationRequired
}

fn verify_candidate(
    input: &Value,
    key_ring: &[OfflineEntitlementSigningKey],
    context: OfflineEntitlementVerificationContext<'_>,
    trusted_time_store: &mut impl TrustedTimeStore,
) -> Option<()> {
    let envelope = envelope_view(input)?;
    let signing_key = key_ring.iter().find(|key| key.key_id == envelope.key_id)?;
    if signing_key.status == OfflineEntitlementSigningKeyStatus::Retired {
        return None;
    }

    let payload_bytes = decode_base64url(envelope.payload_bytes)?;
    let signature_bytes: [u8; 64] = decode_base64url(envelope.signature)?.try_into().ok()?;
    let signature = Signature::from_bytes(&signature_bytes);
    let verifying_key = VerifyingKey::from_bytes(&signing_key.public_key_bytes).ok()?;
    verifying_key
        .verify_strict(&payload_bytes, &signature)
        .ok()?;

    validate_control_plane_document(input).ok()?;
    let claims: OfflineEntitlementClaims = serde_json::from_slice(&payload_bytes).ok()?;

    let issued_at = parse_canonical_utc_seconds(&claims.issued_at)?;
    let expires_at = parse_canonical_utc_seconds(&claims.expires_at)?;
    let envelope_issued_at = parse_canonical_utc_seconds(envelope.issued_at)?;
    let envelope_expires_at = parse_canonical_utc_seconds(envelope.expires_at)?;
    let last_trusted_time = trusted_time_store.read_last_trusted_unix_seconds();

    if last_trusted_time.is_some_and(|last| context.now_unix_seconds < last)
        || context.now_unix_seconds < issued_at
        || context.now_unix_seconds > expires_at
        || expires_at.checked_sub(issued_at)? != OFFLINE_ENTITLEMENT_VALIDITY_SECONDS
        || claims.validity_seconds != OFFLINE_ENTITLEMENT_VALIDITY_SECONDS
        || issued_at != envelope_issued_at
        || expires_at != envelope_expires_at
        || claims.schema_version != "1.0"
        || claims.account_id != context.account_id
        || claims.device_binding != context.device_binding
        || claims.device_binding != envelope.device_binding
        || claims.audience != context.audience
        || claims.audience != envelope.audience
        || claims.entitlement_version != context.entitlement_version
        || issued_at < signing_key.not_before_unix_seconds
        || issued_at > signing_key.not_after_unix_seconds
        || context.now_unix_seconds < signing_key.not_before_unix_seconds
        || context.now_unix_seconds > signing_key.not_after_unix_seconds
    {
        return None;
    }

    trusted_time_store
        .write_last_trusted_unix_seconds(context.now_unix_seconds)
        .ok()
}

pub fn verify_offline_entitlement(
    input: &Value,
    key_ring: &[OfflineEntitlementSigningKey],
    context: OfflineEntitlementVerificationContext<'_>,
    trusted_time_store: &mut impl TrustedTimeStore,
) -> OfflineEntitlementVerdict {
    if verify_candidate(input, key_ring, context, trusted_time_store).is_some() {
        OfflineEntitlementVerdict::Verified
    } else {
        verification_required()
    }
}

#[cfg(test)]
mod tests {
    use super::{OFFLINE_ENTITLEMENT_VALIDITY_SECONDS, parse_canonical_utc_seconds};

    #[test]
    fn canonical_utc_parser_preserves_exact_seven_day_math() {
        let issued = parse_canonical_utc_seconds("2026-08-04T12:00:00.000Z").expect("issued");
        let expires = parse_canonical_utc_seconds("2026-08-11T12:00:00.000Z").expect("expires");
        assert_eq!(expires - issued, OFFLINE_ENTITLEMENT_VALIDITY_SECONDS);
        assert!(parse_canonical_utc_seconds("2026-02-29T12:00:00.000Z").is_none());
        assert!(parse_canonical_utc_seconds("2026-08-04T12:00:00Z").is_none());
    }
}
